### PR TITLE
Text area box

### PR DESCRIPTION
**Pull request by spireken**

### CODE REVIEW

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 1]
```diff
@@ -0,0 +1,634 @@
...
+    &TextAreaBox::on_text_changed);
+  m_placeholder = new ElidedLabel(this);
+  m_placeholder->setFixedSize(0, 0);
+  m_stacked_widget->addWidget(m_placeholder);
+  m_placeholder->setCursor(m_text_edit->cursor());
+  m_placeholder->setAttribute(Qt::WA_TransparentForMouseEvents);
+  m_scroll_box = new ScrollBox(m_stacked_widget, this);
+  m_scroll_box->set(ScrollBox::DisplayPolicy::NEVER,
+    ScrollBox::DisplayPolicy::ON_OVERFLOW);
+  m_scroll_box->get_vertical_scroll_bar().adjustSize();
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:202]: What is the purpose of this ```adjustSize``` call?

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 2]
```diff
@@ -0,0 +1,634 @@
...
+  m_scroll_box->setSizePolicy(QSizePolicy::Minimum, QSizePolicy::Minimum);
+  m_scroll_box->installEventFilter(this);
+  m_scroll_box->findChild<ScrollableLayer*>()->installEventFilter(this);
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:205]: This is almost certainly not a good idea, what is the reason for this?

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 3]
```diff
@@ -0,0 +1,634 @@
...
+  }
+  QWidget::changeEvent(event);
+}
+
+bool TextAreaBox::eventFilter(QObject* watched, QEvent* event) {
+  if(watched == m_text_edit && event->type() == QEvent::FocusOut) {
+    m_submission = m_model->get_current();
+    m_submit_signal(m_submission);
+  } else if(event->type() == QEvent::Wheel) {
+    if(m_text_edit->isReadOnly()) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:274]: Why is wheeling disabled on read-only?

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 4]
```diff
@@ -0,0 +1,634 @@
...
+      return true;
+    }
+  }
+  return QWidget::eventFilter(watched, event);
+}
+
+void TextAreaBox::mousePressEvent(QMouseEvent* event) {
+  m_text_edit->setFocus();
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:282]: Why is this needed? What happens if it's removed? Clicking on a widget should always give it focus by default.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 5]
```diff
@@ -0,0 +1,634 @@
...
+void TextAreaBox::commit_placeholder_style() {
+  auto alignment = m_placeholder_styles.m_alignment.value_or(
+    Qt::Alignment(Qt::AlignmentFlag::AlignLeft));
+  m_placeholder->set_alignment(alignment);
+  auto font = m_placeholder_styles.m_font.value_or(QFont());
+  if(m_placeholder_styles.m_size) {
+    font.setPixelSize(*m_placeholder_styles.m_size);
+  }
+  m_placeholder->setFont(font);
+  m_placeholder->set_text_color(m_placeholder_styles.m_color.value_or(
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:300]: What is the purpose of using an optional for color if the default color is ```QColor```? Better off using ```QColor``` and remove the use of optional.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 6]
```diff
@@ -0,0 +1,634 @@
...
+void TextAreaBox::resizeEvent(QResizeEvent* event) {
+  update_layout();
+  QWidget::resizeEvent(event);
+}
+
+void TextAreaBox::commit_placeholder_style() {
+  auto alignment = m_placeholder_styles.m_alignment.value_or(
+    Qt::Alignment(Qt::AlignmentFlag::AlignLeft));
+  m_placeholder->set_alignment(alignment);
+  auto font = m_placeholder_styles.m_font.value_or(QFont());
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:295]: See below about color and ```optional```.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 7]
```diff
@@ -0,0 +1,634 @@
...
+  update_placeholder_text();
+}
+
+void TextAreaBox::commit_style() {
+  auto stylesheet = QString(
+    R"(QTextEdit {
+      background: transparent;
+      border-width: 0px;)");
+  m_text_edit_styles.m_styles.write(stylesheet);
+  auto alignment = m_text_edit_styles.m_alignment.value_or(
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:311]: See above about placeholder color. If the default is ```AlignLeft``` then use that as ```m_text_edit_styles.m_alignment``` instead of an empty ```optional```.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 8]
```diff
@@ -0,0 +1,634 @@
...
+}
+
+void TextAreaBox::update_display_text() {
+  if(is_read_only()) {
+    if(m_text_edit->toPlainText() != m_model->get_current()) {
+      m_text_edit->blockSignals(true);
+      m_text_edit->setText(m_model->get_current());
+      m_text_edit->blockSignals(false);
+    }
+    auto line_count = std::floor(static_cast<double>((height() -
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:355]: Break up at higher AST depth.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 9]
```diff
@@ -0,0 +1,634 @@
...
+      get_padding_size().height())) /
+      static_cast<double>(m_computed_line_height));
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:357]: Only cast one value to a double, not both.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 10]
```diff
@@ -0,0 +1,634 @@
...
+
+void TextAreaBox::update_display_text() {
+  if(is_read_only()) {
+    if(m_text_edit->toPlainText() != m_model->get_current()) {
+      m_text_edit->blockSignals(true);
+      m_text_edit->setText(m_model->get_current());
+      m_text_edit->blockSignals(false);
+    }
+    auto line_count = std::floor(static_cast<double>((height() -
+      get_padding_size().height())) /
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:356]: Remove excessive parenthesis.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 11]
```diff
@@ -0,0 +1,634 @@
...
+      static_cast<double>(m_computed_line_height));
+    if(line_count > 0) {
+      auto lines = [&] {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:359]: Remove the IIFE, it's not serving any purpose, line 360 completely invalidates the use of an IIFE.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 12]
```diff
@@ -0,0 +1,634 @@
...
+        auto lines = QStringList();
+        for(auto i = 0; i < m_text_edit->document()->blockCount(); ++ i) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:361]: No space after ```++```.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 13]
```diff
@@ -0,0 +1,634 @@
...
+          auto block = m_text_edit->document()->findBlockByNumber(i);
+          if(block.isValid()) {
+            auto block_text = block.text();
+            for(int i = 0; i != block.layout()->lineCount(); ++i) {
+              auto line = block.layout()->lineAt(i);
+              lines.append(block_text.mid(line.textStart(), line.textLength()));
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:367]: 80 column max.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 14]
```diff
@@ -0,0 +1,634 @@
...
+            }
+            if(lines.count() > line_count) {
+              break;
+            }
+          }
+        }
+        return lines;
+      }();
+      if(lines.count() > line_count) {
+        auto is_elided = lines.count() > line_count;
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:380]: This is fishy business, line 379 checks ```lines.count() > line_count```, and then line 380 computes this again, and then line 381 yet again. Fix this whole section up.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 15]
```diff
@@ -0,0 +1,634 @@
...
+    get_padding_size());
+  m_stacked_widget->adjustSize();
+  update_display_text();
+  update_placeholder_text();
+}
+
+void TextAreaBox::update_line_height() {
+  m_computed_line_height =
+    static_cast<int>(static_cast<double>(m_text_edit->font().pixelSize()) *
+    *m_text_edit_styles.m_line_height);
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:439]: If ```m_line_height``` is already a ```double```, then don't cast the ```pixelSize()``` expression to a ```double``` as well.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 16]
```diff
@@ -0,0 +1,634 @@
...
+  }
+}
+
+void TextAreaBox::update_text_alignment() {
+  auto alignment = m_text_edit_styles.m_alignment.value_or(Qt::AlignLeft);
+  auto cursor_pos = m_text_edit->textCursor().position();
+  for(auto i = 0; i < m_text_edit->document()->blockCount(); ++i) {
+    auto block = m_text_edit->document()->findBlockByNumber(i);
+    if(block.isValid()) {
+      auto cursor = m_text_edit->textCursor();
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:464]: This looks very similar to some code written above. See if you can extract the common code.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 17]
```diff
@@ -0,0 +1,634 @@
...
+  cursor.setPosition(cursor_pos);
+  m_text_edit->setTextCursor(cursor);
+}
+
+void TextAreaBox::update_text_edit_width() {
+  auto border_size = get_border_size();
+  auto padding_size = get_padding_size();
+  if(m_text_edit->document()->size().toSize().height() > height() -
+      padding_size.height() - border_size.height() &&
+      !m_text_edit->isReadOnly()) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:480]: When you have an ```A && B``` that share no dependencies, put the cheaper operation on the left-hand side so that the expression can short circuit without doing unnecessary work. ```foo.is_read_only``` is trivial and cheap and if it's not true then you can avoid all this calculation as opposed to the way it's structured now where a bunch of calculations are performed, and then when foo.is_read_only is evaluated, all of those calculations go to waste.

**Review by kranar — CHANGES_REQUESTED** [review]
A lot of excellent work done, but still quite a bit more work needed. Keep it up.

File: Applications/Spire/Include/Spire/Ui/ScrollBox.hpp [hunk 18]
```diff
@@ -66,6 +66,12 @@ namespace Spire {
       /** Sets the horizontal and vertical display policies. */
       void set(DisplayPolicy horizontal_policy, DisplayPolicy vertical_policy);
 
+      /** Returns the vertical ScrollBar. */
+      ScrollBar& get_vertical_scroll_bar();
```
[the remark this answers is not in the document]
Comment (spireken) [reply Applications/Spire/Include/Spire/Ui/ScrollBox.hpp:70]: How can I scroll the ScrollBox to show the cursor without an accessor for the ScrollBars?

[the remark this answers is not in the document]
Comment (kranar) [reply Applications/Spire/Include/Spire/Ui/ScrollBox.hpp:70]: Okay fine, work on everything else and then we can figure this one out.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp (hunk 1 above)
Comment (spireken) [reply Applications/Spire/Source/Ui/TextAreaBox.cpp:202]: In update_text_edit_width(), it's possible that the scroll bar will be shown after the text edit's width is updated, but it's not currently visible. This adjustSize call makes sure that when width() is called that it returns the size the scroll bar would be if it was shown.

On reviewing it, I can move this call to right before the width is queried.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp (hunk 3 above)
Comment (spireken) [reply Applications/Spire/Source/Ui/TextAreaBox.cpp:274]: I don't adjust the size of the text edit when it's read-only so it's possible to scroll it. I can probably clean this up and remove the call to findChild in the above comment by resizing the text edit when it's read-only.

Comment (kranar) [reply Applications/Spire/Source/Ui/TextAreaBox.cpp:274]: Good, clean it up.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp (hunk 4 above)
Comment (spireken) [reply Applications/Spire/Source/Ui/TextAreaBox.cpp:282]: Even with the focus proxy set the text edit isn't receiving focus when the TextAreabox is clicked. I'll dig into it a little more.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp (hunk 1 above)
Comment (kranar) [reply Applications/Spire/Source/Ui/TextAreaBox.cpp:202]: If you want what the width would or "should" be, use the component's size hint. ```adjustSize``` is a very expensive call that produces size effects, ```sizeHint``` produces no side effects and is generally an ```O(1)``` operation.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp (hunk 4 above)
Comment (kranar) [reply Applications/Spire/Source/Ui/TextAreaBox.cpp:282]: Sure, if there's a bug then create a task for it. Don't work around bugs without first reporting them, it results in everyone coming up with hacky ways of working around issues instead of solving it at the root.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 19]
```diff
@@ -0,0 +1,634 @@
...
+          m_text_edit->fontMetrics().horizontalAdvance(ELLIPSES_CHAR));
+        if(is_elided && !last_line.endsWith(ELLIPSES_CHAR)) {
+          last_line.append(ELLIPSES_CHAR);
+        }
+        m_text_edit->blockSignals(true);
+        m_text_edit->setText(lines.join(""));
+        m_text_edit->blockSignals(false);
+      }
+    }
+    update_text_alignment();
```
Comment (shaobinspire) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:397]: Line 397- 399 could be removed, and Line 403 - 404 can do it.

**Review by shaobinspire — CHANGES_REQUESTED** [review]
When“abcd” is input into the TextAreaBox, the cursor can’t be put between ‘c’ and ‘d’ by the mouse, but it can be done using the arrow key.

**Comment by spireken** [comment]
These commits don't address the ScrollBox's ScrollBar getters, nor the issue @shaobinspire pointed out which I think is related to the ScrollableLayer.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 20]
```diff
@@ -0,0 +1,630 @@
...
+  update_layout();
+}
+
+connection TextAreaBox::connect_submit_signal(
+    const SubmitSignal::slot_type& slot) const {
+  return m_submit_signal.connect(slot);
+}
+
+QSize TextAreaBox::sizeHint() const {
+  return m_scroll_box->sizeHint();
```
Comment (shaobinspire) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:275]: I think the sizeHint function is not necessary, because the m_scroll_box is placed in the layout of the TextAreaBox and Qt can handle it automatically.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 21]
```diff
@@ -0,0 +1,630 @@
...
+      static_cast<double>(m_computed_line_height));
+    auto is_elided = false;
+    if(line_count > 0) {
+      auto lines = QStringList();
+      for(auto i = 0; i < m_text_edit->document()->blockCount() && !is_elided;
+          ++i) {
+        auto block = m_text_edit->document()->findBlockByNumber(i);
+        if(block.isValid()) {
+          auto block_text = block.text();
+          for(int i = 0; i != block.layout()->lineCount(); ++i) {
```
Comment (shaobinspire) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:381]: The inner loop and the outer loop have the same iterator name "i", even though it is allowed by the standard, I don't think it's a good practice.

**Review by shaobinspire — CHANGES_REQUESTED** [review]

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 22]
```diff
@@ -0,0 +1,630 @@
...
+#include <QTextBlock>
+#include <QTextDocument>
+#include "Spire/Spire/Dimensions.hpp"
+#include "Spire/Spire/LocalValueModel.hpp"
+#include "Spire/Ui/ScrollableLayer.hpp"
+#include "Spire/Ui/ScrollBar.hpp"
+#include "Spire/Ui/ScrollBox.hpp"
+
+using namespace boost;
+using namespace boost::posix_time;
```
Comment (shaobinspire) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:14]: I don't see where the namespace is used.

**Comment by spireken** [comment]
@shaobinspire There was an update to the ScrollableLayer in spire2018 that fixed the issue you pointing out with selecting the right side text. It's merged into this branch now.

**Review by shaobinspire — APPROVED** [review]
Looks good.

File: Applications/Spire/Include/Spire/Ui/TextAreaBox.hpp [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+      ElidedLabel* m_placeholder;
+      StyleProperties m_placeholder_styles;
+      ScrollBox* m_scroll_box;
+      boost::signals2::scoped_connection m_current_connection;
+      QString m_submission;
+      QString m_placeholder_text;
+      int m_computed_line_height;
+
+      void apply_block_formatting(
+        const std::function<void(const QTextBlock& block)> formatting);
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Ui/TextAreaBox.hpp:103]: Missing ```&```.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 24]
```diff
@@ -0,0 +1,627 @@
...
+    style.get(Placeholder()).set(TextColor(QColor::fromRgb(0xA0, 0xA0, 0xA0)));
+    style.get(Disabled() / Placeholder()).
+      set(TextColor(QColor::fromRgb(0xC8, 0xC8, 0xC8)));
+    return style;
+  }
+}
+
+class TextAreaBox::ContentSizedTextEdit : public QTextEdit {
+  public:
+
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:57]: Remove vertical space.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 25]
```diff
@@ -0,0 +1,627 @@
...
+      set(BackgroundColor(QColor::fromRgb(0, 0, 0, 0))).
+      set(border_color(QColor::fromRgb(0, 0, 0, 0)));
+    style.get(Placeholder()).set(TextColor(QColor::fromRgb(0xA0, 0xA0, 0xA0)));
+    style.get(Disabled() / Placeholder()).
+      set(TextColor(QColor::fromRgb(0xC8, 0xC8, 0xC8)));
+    return style;
+  }
+}
+
+class TextAreaBox::ContentSizedTextEdit : public QTextEdit {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:55]: Why is a ```QTextEdit``` used instead of a ```QPlainTextEdit```?

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 26]
```diff
@@ -0,0 +1,627 @@
...
+    QSize minimumSizeHint() const override {
+      return QSize();
+    }
+
+  private:
+    int m_longest_line_width;
+
+    void on_text_changed() {
+      m_longest_line_width = get_longest_line_width();
+      updateGeometry();
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:94]: What if nothing has changed? Calling ```updateGeometry``` is not cheap.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 27]
```diff
@@ -0,0 +1,627 @@
...
+            std::max(longest, fontMetrics().horizontalAdvance(block.text()));
+        }
+      }
+      return longest;
+    }
+};
+
+class TextAreaBox::ElidedLabel : public QWidget {
+  public:
+
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:112]: Vertical white space.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 28]
```diff
@@ -0,0 +1,627 @@
...
+    explicit ElidedLabel(QWidget *parent = nullptr)
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:113]: Use an inheriting constructor.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 29]
```diff
@@ -0,0 +1,627 @@
...
+      update();
+    }
+
+    void set_line_height(int line_height) {
+      m_line_height = line_height;
+      update();
+    }
+
+    QSize sizeHint() const override {
+      return {width(), height()};
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:137]: This looks unusual, can you confirm it's needed?

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 30]
```diff
@@ -0,0 +1,627 @@
...
+    }
+
+  protected:
+    void paintEvent(QPaintEvent *event) override {
+      auto painter = QPainter(this);
+      painter.setPen(m_text_color);
+      painter.setFont(font());
+      auto fontMetrics = painter.fontMetrics();
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:145]: Why is a copy of this being made?

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 31]
```diff
@@ -0,0 +1,627 @@
...
+      auto y = m_line_height - fontMetrics.height();
+      auto textLayout = QTextLayout(m_text, painter.font());
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:147]: Why is a ```QTextLayout``` constantly being constructed on every paint operation?

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 32]
```diff
@@ -0,0 +1,627 @@
...
+      auto opt = textLayout.textOption();
+      opt.setAlignment(m_alignment);
+      textLayout.setTextOption(opt);
+      textLayout.beginLayout();
+      auto line = textLayout.createLine();
+      while(line.isValid()) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:153]: This appears like a for loop:

    for(auto line = foo(); line.isValid(); line = foo()) { ... }

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 33]
```diff
@@ -0,0 +1,627 @@
...
+  update_placeholder_text();
+}
+
+bool TextAreaBox::is_read_only() const {
+  return m_text_edit->isReadOnly();
+}
+
+void TextAreaBox::set_read_only(bool read_only) {
+  m_text_edit->setReadOnly(read_only);
+  m_scroll_box->get_vertical_scroll_bar().set_position(0);
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:257]: Why does setting read only always reposition the scrollbar?

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 34]
```diff
@@ -0,0 +1,627 @@
...
+  m_placeholder_text = value;
+  update_placeholder_text();
+}
+
+bool TextAreaBox::is_read_only() const {
+  return m_text_edit->isReadOnly();
+}
+
+void TextAreaBox::set_read_only(bool read_only) {
+  m_text_edit->setReadOnly(read_only);
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:256]: This should be a no-op when no change is made.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 35]
```diff
@@ -0,0 +1,627 @@
...
+  }
+  update_text_alignment();
+  update_document_line_height();
+}
+
+void TextAreaBox::update_document_line_height() {
+  apply_block_formatting([=] (const auto& block) {
+    auto cursor = m_text_edit->textCursor();
+    cursor.setPosition(block.position());
+    m_text_edit->setTextCursor(cursor);
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:419]: Why is ```setTextCursor``` called twice without any intermediate side effects on ```m_text_edit```?

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 36]
```diff
@@ -0,0 +1,627 @@
...
+    m_stacked_widget->adjustSize();
+    m_placeholder->show();
+  } else {
+    m_placeholder->hide();
+  }
+}
+
+void TextAreaBox::update_text_alignment() {
+  apply_block_formatting(
+    [=, alignment = m_text_edit_styles.m_alignment] (const auto& block) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:460]: ```alignment``` is never used.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp [hunk 37]
```diff
@@ -0,0 +1,627 @@
...
+      m_scroll_box->get_vertical_scroll_bar().sizeHint().width() -
+      border_size.width() - padding_size.width());
+  } else {
+    m_text_edit->setFixedWidth(width() - border_size.width() -
+      padding_size.width());
+  }
+}
+
+void TextAreaBox::on_current(const QString& current) {
+  if(m_text_edit->toPlainText() != current) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TextAreaBox.cpp:484]: What does this check do? It's not cheap comparing two strings, by itself it's an ```O(N)``` and when you consider that typing happens character by character this makes typing an ```O(N^2)``` operation.

**Review by kranar — CHANGES_REQUESTED** [review]
There's definitely a lot of good work here, but needs a lot more attention to details.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp (hunk 24 above)
Comment (spireken) [reply Applications/Spire/Source/Ui/TextAreaBox.cpp:57]: Do you mean all whitespace between the method definitions?

Comment (kranar) [reply Applications/Spire/Source/Ui/TextAreaBox.cpp:57]: Just line 57.

File: Applications/Spire/Source/Ui/TextAreaBox.cpp (hunk 25 above)
Comment (spireken) [reply Applications/Spire/Source/Ui/TextAreaBox.cpp:55]: The QPlaintTextEdit doesn't support alignments, so I used the QTextEdit so I could reuse setAlignment.

**Review by kranar — CHANGES_REQUESTED** [review]
Looks better, there's a lot of stuff that is potentially uninitialized and the code isn't structured in a way that makes it clear that all fields are guaranteed to be initialized before being read. Can you guarantee this or in cases where no such guarantee exists, initialize fields to a default value?
